neuromation/api/object_storage.py

```diff
@@ -0,0 +1,401 @@
+
+ProgressQueueItem = Optional[StorageProgressEvent]
+
+
+# We extend from FileStatus to make sure our formatting system from Storage can be
+# reused
+@dataclass(frozen=True)
+class ObjStatus(FileStatus):
+
+    permission: Action = Action.READ
+
+    bucket_name: str = ""
+
+    @property
+    def uri(self) -> URL:
+        return _format_bucket_uri(self.bucket_name, self.path)
+
+
+class ObjectStorage(metaclass=NoPublicConstructor):
+    def __init__(self, core: _Core, config: Config) -> None:
+        self._core = core
+        self._config = config
+        self._file_sem = asyncio.BoundedSemaphore(MAX_OPEN_FILES)
+
+    async def list_buckets(self, *, token: Optional[str] = None) -> List[ObjStatus]:
+        url = self._config.obj_url / "b" / ""
+        auth = await self._config._api_auth()
+
+        async with self._core.request("GET", url, auth=auth) as resp:
+            res = await resp.json()
+            return [_obj_status_from_bucket(bucket) for bucket in res]
+
+    async def list_objects(
+        self, bucket_name: str, prefix: str = "", recursive: bool = False,
+    ) -> List[ObjStatus]:
+        url = self._config.obj_url / "o" / bucket_name
+        auth = await self._config._api_auth()
+
+        query = {"recursive": str(recursive).lower()}
+        if prefix:
+            query["prefix"] = prefix
+        url = url.with_query(query)
+
+        contents: List[ObjStatus] = []
+        common_prefixes: List[ObjStatus] = []
+        while True:
+            async with self._core.request("GET", url, auth=auth) as resp:
+                res = await resp.json()
+                contents.extend(
+                    [_obj_status_from_key(bucket_name, key) for key in res["contents"]]
+                )
+                common_prefixes.extend(
+                    [
+                        _obj_status_from_prefix(bucket_name, prefix)
+                        for prefix in res["common_prefixes"]
+                    ]
+                )
+                if res["is_truncated"] and res["contents"]:
+                    start_after = res["contents"][-1]["key"]
+                    url = url.with_query(start_after=start_after)
+                else:
+                    break
+        return common_prefixes + contents
+
+    async def glob_objects(self, bucket_name: str, pattern: str) -> List[ObjStatus]:
+        pattern = pattern.lstrip("/")
+        parts = pattern.split("/")
+        # Limit the search to prefix of keys
+        prefix = ""
+        for part in parts:
+            if _has_magic(part):
+                break
+            else:
+                prefix += part + "/"
+
+        match = re.compile(fnmatch.translate(pattern)).fullmatch
+        res = []
+        for obj_status in await self.list_objects(bucket_name, prefix=prefix):
+            if match(obj_status.name):
+                res.append(obj_status)
+        return res
+
+    async def head_object(self, bucket_name: str, key: str) -> ObjStatus:
```

Comment:
The APIs here are explicitly getting `bucket_name`. Thus when calling we end up with:
```python
head_object(bucket_name=uri.host, key=uri.path.lstrip('/'))
```
Another way to do it would be to get an URL object and parse it here. Maybe it would be a better API.

Reply:
Or maybe try to do similar signature with Storage with `stat` and `open`

Reply:
The file storage operates with `URL` instances.
Maybe, the same abstraction is good for object storage as well.

Reply:
In this case I can use `object_status.uri`, right?

Reply:
Yes, that is correct.